Create a main application for batched processing

acmiyaguchi opened the pull request:
I've added a main application for processing that's self contained in the `application` folder.

The main processing script is located in `bin/process`, which is similar to the batched processing example. It uses google cloud storage for transferring files between servers and GNU parallel for utilizing resources efficiently.

The `prio_processor staging` command contains the ETL for processing republished prio pings in `moz-fx-data-stage-data`. I've tested this on data proc using the following commands:

```
gcloud dataproc clusters create test-cluster \
    --zone us-east1-b \
    --image-version 1.4 \
    --metadata 'PIP_PACKAGES=click' \
    --service-account dataproc-worker@amiyaguchi-dev.iam.gserviceaccount.com \
    --initialization-actions \
    gs://dataproc-initialization-actions/python/pip-install.sh


gcloud dataproc jobs submit pyspark \
    runner.py \
    --cluster test-cluster  \
    --py-files dist/prio_processor-*.egg \
        -- \
        --date 2019-06-27 \
        --input gs://moz-fx-data-stage-data/telemetry-decoded_gcs-sink-doctype_prio/output \
        --output gs://amiyaguchi-dev/prio_staging/
```

I would like to consolidate the application into the root directory, but there's some trouble with distutils vs setuptools when deploying to dataproc. The `bdist_egg` command doesn't work with distutils, probably because of the extension module.

There are a few things that still need to be implemented here:
* The airflow job that will spin up the dataproc clusters and run the staging job
* The airflow job that spins up the kubernetes cluster and runs the docker application (likely on on large machine with 32+ cores)
* The post-processing job for merging the partitioned aggregates and applying the list of origins to the aggregated bit-vectors

I also haven't run any meaningful stress tests yet, but there are a few tests that test integration. I'm also going to look into the behavior when the incoming data does not match the the current set of private keys (it should skip it and emit some sort of logging message that can be parsed by stackdriver).

acmiyaguchi commented:
@whlach This is should be nearly done -- the ETL and parallelized processing pieces are working locally. The last mile is hooking this up to telemetry-airflow for doing the actual processing portion with the production keys. Any feedback as I finish up?

acmiyaguchi commented on application/docker-compose.yml:
```diff
@@ -0,0 +1,61 @@
+      - BUCKET_INTERNAL_PRIVATE=project-a-private
+      - BUCKET_INTERNAL_SHARED=project-a-shared
+      - BUCKET_EXTERNAL_SHARED=project-b-shared
+      - GCP_CREDENTIALS=/app/.credentials
```
@haroldwoo Is this a realistic method for deploying the prio processor? I'm mounting the google application credentials into the docker container like they are in this integration tests for Prio? One of the first steps is to call an authentication function that runs `gcloud auth activate-service-account --key-file "${GCP_CREDENTIALS}"`.

haroldwoo replied:
Should be okay as long as this is this a private repository.

acmiyaguchi replied:
There aren't any private repositories involved other than the `cloudops-infra`. Mounting the json key file is the easiest way to test that the container works, but I'm not sure how it'll work when running the container in the kubernetes cluster.

haroldwoo replied:
k8s clusters would typically use secrets for this sort of thing.  https://kubernetes.io/docs/concepts/configuration/secret/
If this is a long lived cluster we can add them manually

haroldwoo replied:
also, what are the credentials used for? We may be able to grant the svc account role assigned to the gke cluster the right permissions so the container workload uses them.

acmiyaguchi replied:
It's being using for `gsutil`, in particular `stat`, `rsync`, and `cp`. If there's a simpler way to authenticate using the `gcloud` cli that also works with this development workflow, I can make it part of this PR.

acmiyaguchi replied:
I've renamed GCP_CREDENTIALS to GOOGLE_APPLICATION_CREDENTIALS to be more consistent. If I understand correctly, it seems like this is the way that things are being done now with the GKEPodOperator, where credentials are securely written into the container and the environment variable is set.

wlach commented on prio/cli/commands.py:
```diff
@@ -190,6 +192,8 @@ def verify2(
     name = os.path.basename(input_internal)
     outfile = os.path.join(output, name)
+    if not os.path.exists(output):
+        os.makedirs(output)
```
I think you could simplify this to just:

`os.makedirs(output, exist_ok=True)`

wlach commented on application/tox.ini:
```diff
@@ -0,0 +1,17 @@
+[tox]
```
It is kind of confusing to have two seperate tox.ini/setup.pys, one at the root and one in this application directory.

I would recommend either:

1. Consolidating everything into one bundle.
2. Putting the python-libprio bits into a subdirectory off the root.

Likely (2) makes the most sense at this point.

acmiyaguchi replied:
I agree, but I think this should be left until #59 is merged in to avoid excess merge conflict. This PR is important for python multiprocessing and significantly changes how the wrapper library should be used.

wlach replied:
Ok, as long as there is a plan!

wlach commented on application/README.md:
```diff
@@ -0,0 +1,46 @@
+# The Prio Server Runtime
+
+The application folder contains all of the components that are necessary for
```
I would suggest a more meaningful directory name than "application". Perhaps simply "batch-processor"?

acmiyaguchi replied:
The `application` folder is now `processor`.

acmiyaguchi commented:
@wlach This is ready for another look. I've added documentation for an overview into the application design and I've added the tests to CI.

wlach commented on docs/prio-processor-guide.md:
```diff
@@ -0,0 +1,371 @@
+These libraries are used to build up a series of examples that include
+asynchronous I/O, multi-processing, and container orchestration.
+
+### The `prio` cli tool
```
I think I would incorporate any content here into the command-line reference, and just talk about prio-processor here. The prio command line, again, seems like an part of the implementation of the higher-level processor application. Might be useful to *reference* that document here (in case things go wrong), but I don't think we need to duplicate the content here.

wlach commented on docs/prio-processor-guide.md:
````diff
@@ -0,0 +1,371 @@
+
+# Process all partitions in parallel using a semaphore equal to the core count.
+parallel verify1 :: $(list_partitions)
+```
````
This is pretty hard to follow, and seems to basically be bits of the processor script embedded into the documentation? I think it would be better to just talk about the command and its entry points here, rather than trying to explain how it works internally. Can you maybe just come up with a small example describing how to use the script, assuming that the s3/gcloud buckets are set up correctly?

wlach commented on docs/prio-processor-guide.md:
```diff
@@ -0,0 +1,371 @@
+
+There is support for s3 and gcs storage via
+[`gsutil`](https://cloud.google.com/storage/docs/interoperability).
+
```
Probably want a little more detail about how to tell the tool about what buckets you have configured.

wlach commented on docs/prio-processor-guide.md:
```diff
@@ -0,0 +1,371 @@
+
+### Scheduling
+
+The staging frequency should match the processor job frequency. Both servers
```
What is staging frequency?

wlach commented on docs/prio-processor-guide.md:
````diff
@@ -0,0 +1,371 @@
+```
+
+ 
+## Data Exchange
````
Again, I'd probably farm this section off into a reference document. The content looks useful from the point of view of debugging things when they go wrong, but I don't think knowing the exact partitioning scheme should be necessary from the point of view of setting things up initially.

acmiyaguchi commented:
@wlach I've dropped the documentation from this PR and moved it over to #61. This will unblock deploying the application to DockerHub so we can start running this Airflow and check if this will work with the current Kubernetes setup.
